packages/salesforcedx-vscode-core/src/commands/auth/forceAuthWebLogin.ts

```diff
@@ -22,39 +23,126 @@ import {
 import { ForceAuthLogoutAll } from './forceAuthLogout';
 
-import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
+interface DeviceCodeResponse {
+  user_code: string;
+  device_code: string;
+  interval: number;
+  verification_uri: string;
+}
 
-export class ForceAuthWebLoginExecutor extends SfdxCommandletExecutor<
+export class ForceAuthWebLoginContainerExecutor extends SfdxCommandletExecutor<
   AuthParams
 > {
-  protected showChannelOutput = isSFDXContainerMode();
+  protected showChannelOutput = false;
 
   public build(data: AuthParams): Command {
     const command = new SfdxCommandBuilder().withDescription(
       nls.localize('force_auth_web_login_authorize_org_text')
     );
-    if (isSFDXContainerMode()) {
-      command
-        .withArg('force:auth:device:login')
-        .withLogName('force_auth_device_login');
-    } else {
-      command
-        .withArg('force:auth:web:login')
-        .withLogName('force_auth_web_login');
+
+    command
+      .withArg('force:auth:device:login')
+      .withLogName('force_auth_device_login')
+      .withFlag('--setalias', data.alias)
+      .withFlag('--instanceurl', data.loginUrl)
+      .withArg('--setdefaultusername')
+      .withJson();
+
+    return command.build();
+  }
+
+  public execute(response: ContinueResponse<AuthParams>): void {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+    let deviceCodeReceived = false;
+
+    channelService.streamCommandStartStop(execution);
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(cliResponse => {
+      stdOut += cliResponse.toString();
+
+      if (!deviceCodeReceived) {
+        const authUrl = this.parseAuthUrlFromStdOut(stdOut);
+
+        if (authUrl) {
+          deviceCodeReceived = true;
+          // open the default browser
+          vscode.env.openExternal(vscode.Uri.parse(authUrl, true));
+        }
+      }
+    });
+
+    execution.processExitSubject.subscribe(() => {
+      this.logMetric(execution.command.logName, startTime);
+    });
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private parseAuthUrlFromStdOut(stdOut: string): string | undefined {
+    let authUrl;
+    try {
+      const response = JSON.parse(stdOut) as DeviceCodeResponse;
+      const verificationUrl = response.verification_uri;
+      const userCode = response.user_code;
+
+      if (verificationUrl && userCode) {
+        authUrl = `${verificationUrl}?user_code=${userCode}`;
```

Comment:
Adding the query param allows for the device code to be pre-populated in the form input as the first step for auth.